Add GPS Drivers

EricPedley opened the pull request:
Add GPS drivers for u-blox m8 receiver. To test this, I'm using this main program on EVE and running it with breakpoints in the debugger to see what statements it gets to and for inspecting variables:

```
#include "main.h"
#include "gpio.h"
#include "i2c.h"
#include "gps.h"

void SystemClock_Config(void);

typedef GpsUbxM8I2c GPS;


int main(void)
{
    /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
    HAL_Init();

    /* Configure the system clock */
    SystemClock_Config();

    /* Initialize all configured peripherals */
    MX_GPIO_Init();

    MX_I2C1_Init();

    /* USER CODE BEGIN 2 */
    GPS gps(GPS_RST_GPIO_Port, GPS_RST_Pin, &hi2c1, PVT_MESSAGE);
    gps.Init();
    HAL_Delay(1000);

    int lastITOW = 0;
    while(1) {
        volatile GPS::PollResult res = gps.PollUpdate();    
        auto state = gps.GetState();
        if(state == GPS::State::RESPONSE_READY) {
            UBX_NAV_PVT_PAYLOAD sol = *(UBX_NAV_PVT_PAYLOAD*)gps.GetSolution();            
            volatile int diff = sol.iTOW - lastITOW;
            lastITOW = sol.iTOW;
            if((GPSFixType)sol.fixType == GPSFixType::FIX_3D) {
                HAL_Delay(1000);
            }
            gps.Reset();
        } else {
            HAL_Delay(100);
        }
    }
}



void SystemClock_Config(void) {
    RCC_OscInitTypeDef RCC_OscInitStruct = {0};
    RCC_ClkInitTypeDef RCC_ClkInitStruct = {0};

    RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_HSI;
    RCC_OscInitStruct.HSIState = RCC_HSI_ON;
    RCC_OscInitStruct.HSICalibrationValue = RCC_HSICALIBRATION_DEFAULT;
    RCC_OscInitStruct.Prediv1Source = RCC_PREDIV1_SOURCE_HSE;
    RCC_OscInitStruct.PLL.PLLState = RCC_PLL_ON;
    RCC_OscInitStruct.PLL.PLLSource = RCC_PLLSOURCE_HSI_DIV2;
    RCC_OscInitStruct.PLL.PLLMUL = RCC_PLL_MUL9;
    RCC_OscInitStruct.PLL2.PLL2State = RCC_PLL_NONE;
    if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK) {
        Error_Handler();
    }

    RCC_ClkInitStruct.ClockType = RCC_CLOCKTYPE_HCLK | RCC_CLOCKTYPE_SYSCLK | RCC_CLOCKTYPE_PCLK1 | RCC_CLOCKTYPE_PCLK2;
    RCC_ClkInitStruct.SYSCLKSource = RCC_SYSCLKSOURCE_PLLCLK;
    RCC_ClkInitStruct.AHBCLKDivider = RCC_SYSCLK_DIV1;
    RCC_ClkInitStruct.APB1CLKDivider = RCC_HCLK_DIV1;
    RCC_ClkInitStruct.APB2CLKDivider = RCC_HCLK_DIV1;

    if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_1) != HAL_OK) {
        Error_Handler();
    }

    __HAL_RCC_PLLI2S_ENABLE();
}

void Error_Handler(void) {
    __disable_irq();
    while (1) {
    }
}
```

EricPedley committed:
add WIP gps drivers

EricPedley committed:
rename folder to be more specific

EricPedley committed:
switch to pvt

I tested the ECEF polling on EVE and it didn't work, but PVT did.

EricPedley committed:
remove unnecessary files

EricPedley committed:
fix errors with incomplete removal of old code

EricPedley committed:
move enums into class and add more specific poll results

EricPedley committed:
add verbose comments and more specific enum types

EricPedley committed:
add line to not mess up enum formatting

EricPedley committed:
ran clang-format

EricPedley commented on .clang-format:
```diff
@@ -2,3 +2,4 @@ BasedOnStyle:  Google
 AccessModifierOffset: -2
 ColumnLimit: 200
 IndentWidth: 4
+AllowShortEnumsOnASingleLine: false
```
Without this, it tried to put all my enum values on a single line.

DanielHeEGG replied:
Huh ok, plz push this change to the overview repo

EricPedley replied:
https://github.com/UCI-Rocket-Project/rocket2-overview/pull/1

DanielHeEGG commented on gps_ubxm8_i2c/ubxPacket.cpp:
```diff
@@ -0,0 +1,69 @@
+        payloadLength = newByte;
+    } else if (currentPacketIndex == 3) {
+        payloadLength |= (uint16_t)newByte << 8;
+    } else if (currentPacketIndex >= 4 + payloadLength && currentPacketIndex < 6 + payloadLength) {
```
the order here is a bit confusing, also recommend add extra brackets for clarity, perhaps do
```c
} else if ((currentPacketIndex >= 4) && (currentPacketIndex < 4 + payloadLength)) {
// blah blah
} else if (currentPacketIndex == 4 + payloadLength) {
// blah blah
} else {
//blah blah
}
```

DanielHeEGG commented on gps_ubxm8_i2c/gps.cpp:
```diff
@@ -0,0 +1,117 @@
+        state = GPS::State::POLLING_RESPONSE;
+    }
+
+    if (state == GPS::State::POLLING_RESPONSE) {
```
do we want to check right after we send the request or wait for next cycle?

EricPedley replied:
I think there's no harm in checking immediately, on the off-chance the data is available immediately after sending the request.

DanielHeEGG commented on gps_ubxm8_i2c/gps.cpp:
```diff
@@ -0,0 +1,117 @@
+    state = GPS::State::REQUEST_NOT_SENT;
+}
+
+const GPS::State GPS::getState() { return state; }
```
pascal case function names

DanielHeEGG commented on gps_ubxm8_i2c/gps.h:
```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include <i2c.h>
```
i2c.h is not guaranteed to exist. reference the HAL library header file instead ex: `stm32f1xx_hal.h`

DanielHeEGG commented on gps_ubxm8_i2c/gps.h:
```diff
@@ -0,0 +1,57 @@
+ * The most relevant parts are "UBX Protocol" (section 32) and "DDC Port" (section 15).
+ *
+ */
+class GPS {
```
let's stick with the long device name

DanielHeEGG commented on gps_ubxm8_i2c/gps.h:
````diff
@@ -0,0 +1,57 @@
+ *     }
+ * ```
+ *
+ * After a packet is received, you _need_ to call the reset method.
````
add a Init() method to conform to same format as other drivers, the driver should handle the reset pin.
Also, the GPS module is has an interrupt pin that is configurable - may be able to replace checking the buffer size over and over and save time

EricPedley replied:
Why have an `Init()` method instead of just putting that functionality in the constructor?

DanielHeEGG replied:
The instance of the object is defined in the global scope, hence the constructor is executed before `main`. If the constructor contains code that requires interfacing with hardware - it will error out because the hardware isn't configured yet

DanielHeEGG commented on gps_ubxm8_i2c/ubxHelpers.cpp:
```diff
@@ -0,0 +1,34 @@
+
+static uint8_t magicBytes[2] = {0xB5, 0x62};
+
+bool sendUBX(uint8_t* message, uint16_t len, I2C_HandleTypeDef* i2c) {
```
can we just include this as a private method inside the gps driver class?

DanielHeEGG commented on gps_ubxm8_i2c/ubxHelpers.h:
```diff
@@ -0,0 +1,5 @@
+#include "i2c.h"
+#include "ubxMessages.h"
+
+bool sendUBX(uint8_t* message, uint16_t len, I2C_HandleTypeDef* i2c);
```
javadoc-style comments for functions plz

DanielHeEGG commented on gps_ubxm8_i2c/gps.cpp:
```diff
@@ -0,0 +1,117 @@
+        uint16_t dataLen = 0;
+        uint8_t buffer[I2C_BUFFER_SIZE];
+
+        HAL_StatusTypeDef dataLenReadStatus = HAL_I2C_Mem_Read(i2c, 0x42 << 1, 0xFD, 1, lenBytes, 2, 100);
```
make the timeout configurable via macros, see altimeter drivers for example

DanielHeEGG commented on gps_ubxm8_i2c/gps.cpp:
```diff
@@ -0,0 +1,117 @@
+            return GPS::PollResult::DATA_RECEIVE_I2C_FAILED;
+        }
+
+        if (packetReader.isInProgress()) {
```
how does this section handle having potentially multiple ubx packets in one I2C read operation? it seems to return immediately after parsing the first packet and the rest of the buffer is thrown away

EricPedley replied:
yeah because rn we're only ever sending 1 command at once, and we always wait for the end of one packet before sending another. I think handling multiple packets at a time rn is premature optimization.

DanielHeEGG commented on gps_ubxm8_i2c/gps.cpp:
```diff
@@ -0,0 +1,117 @@
+            return GPS::PollResult::NO_DATA;
+        }
+
+        if (dataLen > I2C_BUFFER_SIZE) {
```
I think if the gps module has more than `I2C_BUFFER_SIZE` bytes of data available, we should still parse it during one execution of `pollUpdate()` and not wait for the next cycle

EricPedley replied:
I don't think we should do this because this risks blocking for a long time. Maybe a compromise could be reading at most a fixed multiple of I2C_BUFFER_SIZE like 5 times? Lmk if that sounds good.

DanielHeEGG replied:
I see, I didn't consider that. Let's just leave it as is.

DanielHeEGG commented on gps_ubxm8_i2c/gps.cpp:
```diff
@@ -0,0 +1,117 @@
+
+        if (packetReader.isInProgress()) {
+            for (uint16_t i = 0; i < dataLen; i++) {
+                UBXPacketUpdateResult res = packetReader.update(buffer[i]);
```
i think the logic here and in `UBXPacketReader` can be cleaned up a bit but I have to think about it some more

DanielHeEGG commented on gps_ubxm8_i2c/gps.cpp:
```diff
@@ -0,0 +1,117 @@
+    return GPS::PollResult::POLL_ALREADY_FINISHED;
+}
+
+/**
```
use javadoc/doxygen style comments w/ `@brief` `@param` and `@retval`
VSCode automatically parses them and makes the mouse hover popup a lot nicer

also by convention move the general method comments to header file

DanielHeEGG commented on gps_ubxm8_i2c/gps.cpp:
```diff
@@ -0,0 +1,117 @@
+#include "gps.h"
```
we're generally following google's c++ styleguide so c++ source should be `.cc`

EricPedley committed:
refactor conditional logic in ubxPacket

EricPedley committed:
rename files to .cc

EricPedley committed:
change functions to PascalCase

EricPedley committed:
put enum definition before reference

EricPedley committed:
replace i2c header include

EricPedley committed:
rename gps class to include full info and pull reset pin in constructor

EricPedley committed:
move ubxHelpers to private methods of gps class

EricPedley committed:
add formatting to comments

EricPedley committed:
refactor i2c timeout to a macro

DanielHeEGG commented:
Oh and could you add the ECEF conversion code, after that I think we're good to go

EricPedley committed:
refactor: move private helper to end of file

EricPedley committed:
move pin reset to Init method

EricPedley committed:
fix comments

EricPedley committed:
add newline

EricPedley committed:
decrease column limit

without doing this, it tried putting `Init` in gps.cc on a single line, which was really dumb.

EricPedley committed:
make private variables prefixed with underscore

EricPedley committed:
make reset pin and port private members

EricPedley committed:
replace hard-coded i2c header with macros

EricPedley committed:
succumb to function squashing

removing args form Init made it shorter than 120 and I just gave up on trying to make it not squash

EricPedley commented:
How am I supposed to do ECEF conversion if we can't do floating point arithmetic on the chip? Do we want to do some scuffed stuff with fixed-point approximations or can we just leave it in lat/lng?

EricPedley committed:
add ecef payload back to messages

EricPedley committed:
un-hardcode what message the GPS is sending.

EricPedley committed:
fix syntax errors from not building earlier

EricPedley committed:
add extra parens

EricPedley committed:
add ecef conversion helper

EricPedley committed:
un-const messages

EricPedley committed:
start working on coordinate conversion

EricPedley committed:
add datesheet links to message structs

EricPedley committed:
convert coords to ECEF

EricPedley committed:
fix whitespace problem

EricPedley commented:
@DanielHeEGG can u look at this again? You didn't request changes last time so there's no `request re-review` button for me RN.

EricPedley committed:
fix math includes

DanielHeEGG commented on gps_ubxm8_i2c/coordHelpers.cc:
```diff
@@ -0,0 +1,39 @@
+double Rad2Deg(double rad) { return rad * 180.0 / M_PI; }
+
+// Convert GPS coordinates (latitude, longitude, altitude) to ECEF coordinates
+void GPS2ECEF(double lat, double lon, double alt, double& x, double& y, double& z) {
```
let's make sure it's pascal case, even for acronyms `Gps2Ecef`

DanielHeEGG commented on gps_ubxm8_i2c/gps.cc:
```diff
@@ -0,0 +1,186 @@
+UCIRP_GPS_PAYLOAD ConvertPayloadToECEF(UBX_NAV_PVT_PAYLOAD pvtPayload) {
+    UCIRP_GPS_PAYLOAD payload;
+    payload.gpsFix = pvtPayload.fixType;
+    payload.horizontalPositionAcc = (double)pvtPayload.hAcc / 1000;
```
a separate horizontal and vertical position accuracy only makes sense in long-lat, in ECEF there should only be one position accuracy. idk how to convert tho, I don't think we can just take the sqrt(horz^2 + vert^2)

EricPedley replied:
Yeah ur right. I also wasn't really sure how to convert but I figure taking the max of the two is a good conservative way.

DanielHeEGG commented on gps_ubxm8_i2c/coordHelpers.h:
```diff
@@ -0,0 +1,10 @@
+// Convert degrees to radians
```
i know this is only included in one other file, but just in case, include guard?

DanielHeEGG commented on gps_ubxm8_i2c/gps.h:
```diff
@@ -0,0 +1,76 @@
+#if defined(STM32F1)
+#include "stm32f1xx_hal.h"
+#endif
+#if defined(STM32F4XX)
```
`#if defined(STM32F4xx)`

DanielHeEGG replied:
actually u can just copy this block
```c
#if defined(STM32F1)
#include "stm32f1xx_hal.h"
#elif defined(STM32F4xx)
#include "stm32f4xx_hal.h"
#endif
```

DanielHeEGG commented on gps_ubxm8_i2c/gps.h:
```diff
@@ -0,0 +1,76 @@
+#include "stm32f4xx_hal.h"
+#endif
+
+#include <gpio.h>
```
you shouldn't ever need to reference gpio.h or i2c.h, these are project specific and may not exist

DanielHeEGG commented on gps_ubxm8_i2c/ubxPacket.h:
```diff
@@ -0,0 +1,42 @@
+
+#include <stdint.h>
+
+#ifndef PACKET_READER_PAYLOAD_SIZE
```
u can get rid of the `#ifndef` check, but would recommend prefixing all driver macros with `GPS_` etc

DanielHeEGG commented on gps_ubxm8_i2c/gps.h:
```diff
@@ -0,0 +1,76 @@
+ *     while(1) {
+ *        if(myGPS.PollUpdate() == GPS::PollResult::POLL_FINISHED) {
+ *            UBX_NAV_PVT_PAYLOAD pvtData = myGPS.GetSolution();
+ *            myGPS.Reset();
```
include example of new convert to ecef function would be nice

DanielHeEGG commented on gps_ubxm8_i2c/ubxMessages.h:
```diff
@@ -0,0 +1,66 @@
+
+static uint8_t PVT_MESSAGE[4] = {0x01, 0x07, 0x00, 0x00};
+
+// https://content.u-blox.com/sites/default/files/products/documents/u-blox8-M8_ReceiverDescrProtSpec_UBX-13003221.pdf#%5B%7B%22num%22%3A777%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C0%2C654.8%2Cnull%5D
```
clean up url, just `https://content.u-blox.com/sites/default/files/products/documents/u-blox8-M8_ReceiverDescrProtSpec_UBX-13003221.pdf` is a valid link

DanielHeEGG reviewed (changes requested):
looks good! some more minor suggestions, little functional change. Plz fix file namings and ensure snake case. @MichaelKrinsky plz take a look as well

MichaelKrinsky commented on gps_ubxm8_i2c/coordHelpers.cc:
```diff
@@ -0,0 +1,39 @@
+}
+
+// Convert NED (North-East-Down) velocity to ECEF velocity
+void NED2ECEFVel(double lat, double lon, double alt, double vn, double ve, double vd, double& vx, double& vy, double& vz) {
```
Pascal Case

MichaelKrinsky commented:
Looked through everything. Daniel covered everything I found. Once done with his changes should be good.

EricPedley committed:
add include guard on coorhelpers header

EricPedley committed:
pascal case function names

EricPedley committed:
fix macros

EricPedley committed:
rename and dont ifndef packet reader size

EricPedley committed:
shorten datasheet link

EricPedley committed:
add example of ecef conversion function

EricPedley committed:
use only 1 accuracy and fix duplicate structs

EricPedley committed:
fix header order

EricPedley committed:
update example

DanielHeEGG commented on gps_ubxm8_i2c/gps.cc:
```diff
@@ -0,0 +1,186 @@
+ */
+const GpsUbxM8I2c::PollResult GpsUbxM8I2c::PollUpdate() {
+    if (_state == GpsUbxM8I2c::State::REQUEST_NOT_SENT) {
+        sendUBX(_ubxMessage);
```
no checks here for successful command sent

DanielHeEGG commented on gps_ubxm8_i2c/gps.cc:
```diff
@@ -0,0 +1,186 @@
+    double velE = (double)pvtPayload.velE / 1000;
+    double velD = (double)pvtPayload.velD / 1000;
+
+    Ned2EcefVel(lat, lon, alt, velN, velE, velD, payload.ecefVX, payload.ecefVY, payload.ecefVZ);
```
no return value

EricPedley committed:
check ubx send status and actually return payload

EricPedley committed:
swap packet reader conditionals (fixed dumbass bug)

EricPedley committed:
rename files and add missing namespace to payload conversion

EricPedley committed:
make payload conversion pass-by-reference

EricPedley committed:
make conversion function private and just return simpler struct

EricPedley commented:
so we merging this or what? At this point I think any changes are just bugfixes and they make more sense as separate PRs instead of tacking onto this one.
